File: tools/build-occurrence-lga.py
#!/usr/bin/env python3
"""
Resolve every catalogued occurrence to the LGA whose polygon contains it.

Nearest-centroid assignment is wrong often enough to matter — "Maru Gold Belt"
lands in Kaura Namoda because that centroid happens to be closer than Maru's.
Point-in-polygon is the correct test, with nearest-centroid kept only as a
fallback for points that fall just outside every polygon (coastline slivers,
simplified boundaries).

Writes data/occurrence-lga.json:  { "NMI-0001": {"lga": "Anka", "exact": true}, ... }

Run:  python3 tools/build-occurrence-lga.py
"""
import json
import math
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
LGA_DIR = ROOT / 'data' / 'lga'
OUT = ROOT / 'data' / 'occurrence-lga.json'


def load_deposits():
    """Pull the DEPOSITS array out of the ES module without a JS runtime."""
    src = (ROOT / 'js' / 'data' / 'fixtures.js').read_text()
    start = src.index('export const DEPOSITS = [') + len('export const DEPOSITS = ')
    depth, i = 0, start
    while i < len(src):
        if src[i] == '[':
            depth += 1
        elif src[i] == ']':
            depth -= 1
            if depth == 0:
                break
        i += 1
    return json.loads(src[start:i + 1])


def state_codes():
    src = (ROOT / 'js' / 'data' / 'fixtures.js').read_text()
    block = src[src.index('export const STATES = {'):src.index('export const DEPOSITS')]
    out = {}
    for m in re.finditer(r'"([^"]+)":\s*\{\s*"code":\s*"([A-Z]{2})"', block):
        out[m.group(1)] = m.group(2)
    return out


def in_ring(x, y, ring):
    inside = False
    j = len(ring) - 1
    for i in range(len(ring)):
        xi, yi = ring[i][0], ring[i][1]
        xj, yj = ring[j][0], ring[j][1]
        if (yi > y) != (yj > y):
            xint = (xj - xi) * (y - yi) / ((yj - yi) or 1e-15) + xi
            if x < xint:
                inside = not inside
        j = i
    return inside


def contains(geom, x, y):
    polys = [geom['coordinates']] if geom['type'] == 'Polygon' else geom['coordinates']
    for poly in polys:
        if in_ring(x, y, poly[0]) and not any(in_ring(x, y, h) for h in poly[1:]):
            return True
    return False


def main():
    deposits = load_deposits()
    codes = state_codes()
    cache, out = {}, {}
    exact = fallback = missing = 0

    for d in deposits:
        code = codes.get(d['state'])
        path = LGA_DIR / f'{code}.geojson'
        if not code or not path.exists():
            missing += 1
            continue

        if code not in cache:
            cache[code] = json.loads(path.read_text())['features']
        feats = cache[code]

        hit = next((f for f in feats if contains(f['geometry'], d['lng'], d['lat'])), None)
        if hit:
            out[d['id']] = {'lga': hit['properties']['name'], 'exact': True}
            exact += 1
            continue

        # Outside every polygon — snap to the nearest centroid so the record is
        # still usable, but flag it so the UI can say "approximate".
        best, bd = None, 1e18
        for f in feats:
            c = f['properties']['centroid']
            q = (c[0] - d['lat']) ** 2 + (c[1] - d['lng']) ** 2
            if q < bd:
                bd, best = q, f['properties']['name']
        out[d['id']] = {'lga': best, 'exact': False}
        fallback += 1

    OUT.write_text(json.dumps(out, indent=0, sort_keys=True))
    print(f'{len(out)} occurrences  exact={exact} fallback={fallback} missing={missing}')
    print(f'wrote {OUT.relative_to(ROOT)}')


if __name__ == '__main__':
    main()

File: tools/build-occurrence-lga.js
#!/usr/bin/env node
/**
 * Resolve every catalogued occurrence to the LGA whose polygon contains it.
 *
 * Nearest-centroid assignment is wrong often enough to matter — "Maru Gold
 * Belt" lands in Kaura Namoda because that centroid happens to be closer than
 * Maru's. Point-in-polygon is the correct test, with nearest-centroid kept
 * only as a fallback for points that fall just outside every polygon
 * (coastline slivers, simplified boundaries).
 *
 * Writes data/occurrence-lga.json:
 *   { "NMI-0001": { "lga": "Anka", "exact": true }, ... }
 *
 * Run:  node tools/build-occurrence-lga.js
 */

const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const LGA_DIR = path.join(ROOT, 'data', 'lga');
const OUT = path.join(ROOT, 'data', 'occurrence-lga.json');
const FIXTURES = path.join(ROOT, 'js', 'data', 'fixtures.js');

/** Pull an exported array/object literal out of the ES module by brace matching. */
function extractLiteral(src, exportName, open, close) {
  const marker = `export const ${exportName} = `;
  const start = src.indexOf(marker) + marker.length;
  let depth = 0;
  let i = start;
  for (; i < src.length; i += 1) {
    if (src[i] === open) depth += 1;
    else if (src[i] === close) {
      depth -= 1;
      if (depth === 0) break;
    }
  }
  return JSON.parse(src.slice(start, i + 1));
}

function loadDeposits(src) {
  return extractLiteral(src, 'DEPOSITS', '[', ']');
}

/** name -> two-letter code, read straight from the STATES block. */
function stateCodes(src) {
  const block = src.slice(src.indexOf('export const STATES = {'), src.indexOf('export const DEPOSITS'));
  const out = {};
  const re = /"([^"]+)":\s*\{\s*"code":\s*"([A-Z]{2})"/g;
  let m;
  while ((m = re.exec(block)) !== null) out[m[1]] = m[2];
  return out;
}

function inRing(x, y, ring) {
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i, i += 1) {
    const [xi, yi] = ring[i];
    const [xj, yj] = ring[j];
    if ((yi > y) !== (yj > y)) {
      const xint = ((xj - xi) * (y - yi)) / ((yj - yi) || 1e-15) + xi;
      if (x < xint) inside = !inside;
    }
  }
  return inside;
}

function contains(geom, x, y) {
  const polys = geom.type === 'Polygon' ? [geom.coordinates] : geom.coordinates;
  return polys.some((poly) => inRing(x, y, poly[0]) && !poly.slice(1).some((h) => inRing(x, y, h)));
}

function main() {
  const src = fs.readFileSync(FIXTURES, 'utf8');
  const deposits = loadDeposits(src);
  const codes = stateCodes(src);

  const cache = new Map();
  const out = {};
  let exact = 0;
  let fallback = 0;
  let missing = 0;

  for (const d of deposits) {
    const code = codes[d.state];
    const file = path.join(LGA_DIR, `${code}.geojson`);
    if (!code || !fs.existsSync(file)) { missing += 1; continue; }

    if (!cache.has(code)) {
      cache.set(code, JSON.parse(fs.readFileSync(file, 'utf8')).features);
    }
    const feats = cache.get(code);

    const hit = feats.find((f) => contains(f.geometry, d.lng, d.lat));
    if (hit) {
      out[d.id] = { lga: hit.properties.name, exact: true };
      exact += 1;
      continue;
    }

    // Outside every polygon — snap to the nearest centroid so the record is
    // still usable, but flag it so the UI can say "approximate".
    let best = null;
    let bd = Infinity;
    for (const f of feats) {
      const c = f.properties.centroid;
      const q = (c[0] - d.lat) ** 2 + (c[1] - d.lng) ** 2;
      if (q < bd) { bd = q; best = f.properties.name; }
    }
    out[d.id] = { lga: best, exact: false };
    fallback += 1;
  }

  const sorted = Object.fromEntries(Object.keys(out).sort().map((k) => [k, out[k]]));
  fs.writeFileSync(OUT, JSON.stringify(sorted, null, 0));
  console.log(`${Object.keys(out).length} occurrences  exact=${exact} fallback=${fallback} missing=${missing}`);
  console.log(`wrote ${path.relative(ROOT, OUT)}`);
}

main();
